airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,445 @@
+        self._update_current_window()
+
+        if self._calls_num >= self._call_limit:
+            reset_in = self.current_window.end - datetime.datetime.now()
+            error_message = (
+                f"reached maximum number of allowed calls {self._call_limit} " f"per {self._offset} interval, next reset in {reset_in}."
+            )
+            raise CallRateLimitHit(
+                error=error_message,
+                item=request,
+                weight=weight,
+                rate=f"{self._call_limit} per {self._offset}",
+                time_to_wait=reset_in,
+            )
+
+        self._calls_num += 1
+
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Update call rate counters, by default only reacts to decreasing updates of available_calls and changes to call_reset_ts
+
+        :param available_calls:
+        :param call_reset_ts:
+        """
+        self._update_current_window()
+        current_available_calls = self._call_limit - self._calls_num
+
+        if available_calls is not None and current_available_calls > available_calls:
+            logger.info("got update from api, adjusting available calls from %s to %s", current_available_calls, available_calls)
+            self._calls_num = self._call_limit - available_calls
+
+        if call_reset_ts is not None and call_reset_ts != self.current_window.end:
+            logger.info("got update from api, adjusting reset time from %s to %s", self.current_window.end, call_reset_ts)
+            self.current_window.end = call_reset_ts
+
+    def _update_current_window(self) -> None:
+        now = datetime.datetime.now()
+        if now > self.current_window.end:
+            logger.info("current window %s has being reset, %s calls available now", self.current_window, self._call_limit)
+            self._current_window = TimeWindow(self._current_window.end, self._current_window.end + self._offset)
+            self._calls_num = 0
+
+
+class MovingWindowCallRatePolicy(AbstractCallRatePolicy):
+    """
+    Policy to control requests rate implemented on top of PyRateLimiter lib.
+    """
+
+    def __init__(self, rates: list[Rate]):
+        """Constructor
+
+        :param rates: list of rates, the order is important and must be ascending
+        """
+        pyrate_rates = [PyRateRate(limit=rate.limit, interval=int(rate.interval.total_seconds() * 1000)) for rate in rates]
+        self._bucket = InMemoryBucket(pyrate_rates)
+        # Limiter will create the background task that clears old requests in the bucket
+        self._limiter = Limiter(self._bucket)
+
+    def try_acquire(self, request: Any, weight: int) -> None:
+        try:
+            self._limiter.try_acquire(request, weight=weight)
+        except BucketFullException as exc:
+            item = self._limiter.bucket_factory.wrap_item(request, weight)
+            assert isinstance(item, RateItem)
```

Comment:
is there a chance these asserts might ever fail?

Reply:
This is for mypy, a common technique to let it know that we expect this item to match specific type without nesting your code in deep nested `If`s